### DISCUSSION TITLE
Is my idea a good candidate for ML or DL?

### DISCUSSION THREAD

**Discussion by elewis33** [mrdbourke/pytorch-deep-learning#157] [Q&A] [answered]
I've got a significant repository of email messages/threads that I'd like to analyze and essentially create a sort of prediction engine, where I can say, if a specific topic comes up in the future, I should probably reach out to these specific people to either get them involved or answer some questions. So aside from a trained model of some sort, my input would be a current email message and the output would be, based on historical analysis, you should reach out to these specific people based on the topic of the email message.

I know I can do a certain amount of analysis just using a database, but I'm thinking that the nuances of written language in email messages might make this a better candidate for ML/DL. 

Does anyone have any thoughts about this particular problem/model? 

A couple details that might help understand what I'm dealing with. 

- I'm using my personal email archive, so these are only emails that are either addressed directly to me in the from field or I'm CC'd on the message. 
- My email archive is organized into folders that correspond to specific projects, to there is already a sort of labeling/tagging for the messages in a single folder. 

I'd love to hear what people think about this idea because it's been rattling around in my head for a long time and would like to get some of the pros and cons from others that know the ML/DL field better than me. Thank you! 

Earl

**Comment by fivefishstudios** [comment]
I'm still new/learning ML, so this may not be the best answer. There is a website called Hugging Face (https://huggingface.co/) where you can download pre-trained models for your specific problem and maybe just further train them for your specific problem.

Example, here is a model that has been trained to recognize four types of entities/tokens: like location (LOC), organizations (ORG), person (PER) and Miscellaneous (MISC). 

I fed it a sample sentence of "The XYZ project at New York is delayed due to problems in acquiring building materials by Joe, purchaser at ABC Hardware Company. " 

and the model recognized which text are persons, organizations, locations and miscellaneous items. 

[![joe.png](https://i.postimg.cc/VsSRVHBB/joe.png)](https://postimg.cc/14QwgBdf)

 I guess this model or something similar to it can be used as a starting point. Then further link the org/person to your existing database, maybe use SQL LIKE statement for company names and SQL SOUNDEX for person's names.

**Reply by mrdbourke** [reply]
Fantastic use case!

**Reply by elewis33** [reply]
Thanks so much @fivefishstudios! I'm pretty determined to make this into something useful so your suggestions are greatly appreciated. I'll report back if I make any sort of reasonable progress.

**Comment by mrdbourke** [comment, marked as answer]
Hey @elewis33,

Like @fivefishstudios said, there's a great resource at https://huggingface.co with plenty of pretrained natural language processing (NLP)  models.

What you might want to start looking for is a "zero-shot classification model" where you can provide an example text and the model will try to classify the topics based on what's the in the text *with no pre-assigned labels*.

See this example model here: https://huggingface.co/facebook/bart-large-mnli?candidateLabels=mobile%2C+website%2C+billing%2C+account+access&multiClass=false&text=Last+week+I+upgraded+my+iOS+version+and+ever+since+then+my+phone+has+been+overheating+whenever+I+use+your+app. 

After you've tired that, you can go into more of a fine-tuning mode and use your existing folders/archive as the labels for new emails coming through.

So an existing model can be tailored to your own labels.

I'd research something like "how to fine-tune an NLP model huggingface".

But on the whole your use case is ideal for ML as NLP models have drastically improved over the past couple of years.

**Reply by elewis33** [reply]
@mrdbourke thanks so much for the response and suggestions. After I got into huggingface.co it got a little overwhelming with how things are organized so I backed off initially. But will definitely jump in there again to look at your pointer and the suggestions you made. I'm really anxious to see if this is anything. /s (snarky Jerry Seinfeld reference). Thanks again!

**Reply by mrdbourke** [reply]
There's a lot there! And it will take a while to get around everything. But with practice you will likely discover a fair few things that can help you on your journey!
